core/io/pck_reader.cpp

```diff
@@ -0,0 +1,97 @@
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "pck_reader.h"
+
+#include "core/error/error_list.h"
+#include "core/error/error_macros.h"
+#include "core/io/file_access_pack.h"
+#include "core/variant/variant.h"
+
+Error PCKReader::open(const String &p_path, uint64_t p_offset) {
+	if (file.is_open()) {
+		close();
+	}
+
+	return file.try_open_pack(p_path, p_offset) ? OK : FAILED;
+}
+
+Error PCKReader::close() {
+	ERR_FAIL_COND_V_MSG(!file.is_open(), FAILED, "PCKReader cannot be closed because it is not open.");
+
+	file.close();
+	return OK;
+}
+
+PackedStringArray PCKReader::get_files() {
+	ERR_FAIL_COND_V_MSG(!file.is_open(), PackedStringArray(), "PCKReader must be opened before use.");
+
+	PackedStringArray result;
+	for (const Pair<String, PackedData::PackedFile> &pair : file.get_files()) {
+		result.append(pair.first);
+	}
+
+	return result;
+}
+
+PackedByteArray PCKReader::read_file(const String &p_path, bool p_case_sensitive) {
+	ERR_FAIL_COND_V_MSG(!file.is_open(), PackedByteArray(), "PCKReader must be opened before use.");
+
+	Ref<FileAccess> f = file.get_file(p_path, p_case_sensitive);
+	ERR_FAIL_COND_V_MSG(f.is_null(), PackedByteArray(), "Failed to get file from PCK.");
+
+	PackedByteArray res;
+	res.resize(f->get_length());
+	f->get_buffer((uint8_t *)res.ptr(), res.size());
+
+	return res;
+}
+
+bool PCKReader::file_exists(const String &p_path, bool p_case_sensitive) {
+	ERR_FAIL_COND_V_MSG(!file.is_open(), false, "PCKReader must be opened before use.");
+
+	return file.file_exists(p_path, p_case_sensitive);
+}
+
+PCKReader::PCKReader() {}
+
+PCKReader::~PCKReader() {
+	if (file.is_open()) {
+		close();
+	}
+}
+
+void PCKReader::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("open", "path", "offset"), &PCKReader::open);
```

Comment:
Since PCK reader is user accessible, it might be worth adding the encryption key to the `open` arguments (and fall back to the embedded key if it's empty).